src/transactions/OfferExchange.cpp

```diff
@@ -1391,4 +1454,129 @@ convertWithOffers(
+    // - the exchange with the book succeeded so we should use that
+    // - the exchange with the book failed so we should return that error code
+    //   for compatibility with old behavior
+    return true;
+}
+
+ConvertResult
+convertWithOffersAndPools(
+    AbstractLedgerTxn& ltxOuter, Asset const& sheep, int64_t maxSheepSend,
+    int64_t& sheepSend, Asset const& wheat, int64_t maxWheatReceive,
+    int64_t& wheatReceived, RoundingType round,
+    std::function<OfferFilterResult(LedgerTxnEntry const&)> filter,
+    std::vector<ClaimAtom>& offerTrail, int64_t maxOffersToCross)
+{
+    ZoneScoped;
+
+    // If offerTrail is not empty at the start, then the limit maxOffersToCross
+    // will not be imposed correctly.
+    releaseAssertOrThrow(offerTrail.empty());
+
+    sheepSend = 0;
+    wheatReceived = 0;
+
+    // Compute the exchange from the liquidity pool but don't actually do the
+    // exchange
+    std::optional<ExchangedQuantities> poolExchange;
+    {
+        ExchangedQuantities res;
+        if (exchangeWithPool(ltxOuter, sheep, maxSheepSend, res.sheepSend,
+                             wheat, maxWheatReceive, res.wheatReceived, round))
+        {
+            poolExchange = std::make_optional<ExchangedQuantities>(res);
+        }
+    }
+
+    // Maybe use the order book
+    {
+        LedgerTxn ltxConvertWithOffers(ltxOuter);
+
+        ExchangedQuantities bookExchange;
+        auto convertRes = convertWithOffers(
+            ltxConvertWithOffers, sheep, maxSheepSend, bookExchange.sheepSend,
+            wheat, maxWheatReceive, bookExchange.wheatReceived, round, filter,
+            offerTrail, maxOffersToCross);
+
+        if (shouldConvertWithOffers(poolExchange, bookExchange))
+        {
+            sheepSend = bookExchange.sheepSend;
+            wheatReceived = bookExchange.wheatReceived;
+            ltxConvertWithOffers.commit();
+            return convertRes;
+        }
+    }
+
+    // Clear offerTrail because it was initially empty and we didn't actually
+    // cross the order book
+    offerTrail.clear();
+    if (static_cast<uint64_t>(maxOffersToCross) <= 1)
+    {
+        // offerTrail is going to be too long after exchanging with the
+        // liquidity pool
+        return ConvertResult::eCrossedTooMany;
+    }
+
+    // Use the liquidity pool
+    auto poolID = getPoolID(sheep, wheat, LIQUIDITY_POOL_FEE_V18);
+    auto lp = loadLiquidityPool(ltxOuter, poolID);
+    // lp must exist at this point because
+    // - shouldConvertWithOffers returned false
+    // - shouldConvertWithOffers returning false implies poolExchange is not
+    //   empty
+    // - poolExchange being not empty implies that the liquidity pool exists
+    auto cp = [&lp]() -> auto&
+    {
+        return lp.current().data.liquidityPool().body.constantProduct();
+    };
+
+    sheepSend = poolExchange->sheepSend;
+    wheatReceived = poolExchange->wheatReceived;
+    if (sheep == cp().params.assetA && wheat == cp().params.assetB)
```

Comment:
Something bothers me a bit about this wrt `exchangeWithPool` as there are many assumptions baked into this code.
As in: the code depends on computations done in `exchangeWithPool` continuing to be valid up to this point, so why not have `poolExchange` return the `reserveADelta` and `reserveBDelta` instead?

Either that or avoid this caching altogether and
* make the `exchangeWithPool` (the function) actually commit changes by moving the creation of the non-committing `LedgerTxn` to where we call `exchangeWithPool(ltxInner,...`
* call `exchangeWithPool` again at the end of this function, this time keeping the side effects and throwing if it can't be done

Reply:
I'll do one one of the options to avoid the caching altogether, because I think it will be easier to maintain in the future.

But I don't understand you're point about

> the code depends on computations done in `exchangeWithPool` continuing to be valid up to this point, so why not have `poolExchange` return the `reserveADelta` and `reserveBDelta` instead?

I don't see what would be materially different about returning `reserveADelta` and `reserveBDelta` compared to `sheepSend` and `wheatReceived`. You need to match up `assetA/assetB` with `sheep/wheat` no matter what, because you need one to update the reserves and one to produce the `ClaimAtom`.

Reply:
The comment was to avoid rematching assetA/B with sheep/wheat for the sole purpose of updating reserves.
Does not matter anyways as this code goes away if we don't cache anything (either as "diff" like here or the equivalent by returning the actual pool entry in `ExchangedQuantities` which I was alluding to as well).